### DISCUSSION TITLE
Getting "TypeError: Unable to parse some datatype(s) found in args" while executing Application builder

### DISCUSSION THREAD

**Discussion by soumitraghosh388** [AMDResearch/Riallto#57] [General]
I am using 2 kernels : master and slave with n number of instances of slave kernels, where n depends on the input. I have defined graph and am trying to build the application.

Graph code below:

```
class Sample_app(AppBuilder):
    def __init__(self, kernel_1: Kernel, kernel_2: Kernel):
        self.master = kernel_1
        self.slave = kernel_2
        super().__init__()
    def callgraph(self, x_in_1, x_in_2, x_out):
            out_1, out_2 = self.master(x_in_1[0], x_in_2[0])
            slave_out = self.slave (out_1, out_2)
            x_out = [slave_out]
```
  
Creating application builder as mentioned below :          

```
app_builder = Sample_app(kernel_1=master, kernel_2=slave)
x_in_1 = np.array([1028112570], dtype=np.int32)
x_in_2 = np.array([526679170], dtype=np.int32)
x_out = np.array([0], dtype=np.int32)
app_builder(x_in_1, x_in_2, x_out)
```

I am getting below error while executing application builder code :

**TypeError: Unable to parse some datatype(s) found in args: (1028112570, 526679170) ... currently supporting [<class 'npu.build.port.BufferPort'>, <class 'npu.build.buffers.Buffer'>, <class 'numpy.ndarray'>, <class 'int'>] datatypes**

**Comment by mariodruiz** [comment]
Hi @soumitraghosh388,

Would you be able to share the kernels source code?
For the app description, it looks like the `master` kernel takes two input arguments (both scalars) and  the `slave` returns a scalar. Is this correct?

**Reply by soumitraghosh388** [reply]
Hi Mario,

Master and slave definitions are mentioned below:

```
void master(input_stream_int32* __restrict in0, input_stream_int32* __restrict in1, output_stream_int32* __restrict out0, output_stream_int32* __restrict out1);

void slave(input_stream_int32* __restrict in0, input_stream_int32* __restrict in1, output_stream_int32* __restrict out);
```
within master code I am taking inputs as scalars and returning scalars. But in slave kernel I am taking input as scalars as well as vectors and output is vector.

The graph code looks like below (using adf.h):

```
class simpleGraph : public graph {
    private:
        kernel kn[N+1];

    public:
        input_plio p_s0;
        input_plio p_s1;
        output_plio p_s2;

        simpleGraph() {
            p_s0 = input_plio::create("StreamIn0", plio_128_bits, "data/input0.txt");
            p_s1 = input_plio::create("StreamIn1", plio_128_bits, "data/input1.txt");
	    p_s2 = output_plio::create("StreamOut0", plio_128_bits, "output.txt");
            kn[0] = kernel::create(master);
            source(kn[0]) = "master.cc";
            runtime<ratio>(kn[0]) = 0.9;
            
            connect<CONNECTION>(p_s0.out[0], kn[0].in[0]);
            connect<CONNECTION>(p_s1.out[0], kn[0].in[1]);

            for (int i=1; i<=N;i++)
            {
                kn[i] = kernel::create(slave);
                source(kn[i]) = "slave.cc";
                connect<CONNECTION>(kn[i-1].out[0], kn[i].in[0]);
                runtime<ratio>(kn[i]) = 0.9;
            }
          
            for (int i=1; i<=N;i++)
            {
                connect<CONNECTION>(kn[0].out[1], kn[i].in[1]);
            }
	    connect<CONNECTION>(kn[N].out[0], p_s2.in[0]);
        };
};
```

**Reply by mariodruiz** [reply]
Hi @soumitraghosh388,

Unfortunately, in riallto we do not support kernel with streaming interfaces, we only support buffers (arrays).
You will have to convert the kernels to use buffers as described in the notebooks in section 4.

**Reply by soumitraghosh388** [reply]
Hi Mario,

Thanks for the reply. I made required changes to the code as shown below :

**Kernel code (master.cpp) :**

```
void aie_master(int32_t* in0, int32_t* in1, int32_t* out0, int32_t* out1)
{    out0[0] = in0[0];
    out1[0] = in1[0];
}
extern "C" {

void master(int32_t* in0, int32_t* in1, int32_t* out0, int32_t* out1) {
    aie_master(in0, in1, out0, out1);
}
}
```
**Graph class code :**

```
class Graph_vector(AppBuilder):

    def __init__(self, kernel:Kernel):
        self.kernel = kernel
        super().__init__()

    def callgraph(self,  x_in_1:np.ndarray, x_in_2:np.ndarray, x_out:np.ndarray) -> None:
        x_out = self.kernel(x_in_1, x_in_2)
        print(x_out)
```

**Helper class :**

```
class KernelObjCall():
    """ Helper function that generates new kernel objects."""
    def __new__(cls, cpp, bfx, *args):
        kobj = Kernel(cpp, bfx)
        return kobj(*args) if len(args) > 0 else kobj

class Master():
    def __new__(cls, *args):
        cpp = "master.cpp"
        return KernelObjCall(cpp, cls.behavioralfx, *args)
    def behavioralfx(self):
        pass
```

While calling the kernel as shown below :

```
x_in_1 = np.array([1], dtype=np.int32)
x_in_2 = np.array([2], dtype=np.int32)
x_out = np.array([0], dtype=np.int32)
ker_m = Master()
app_builder = Graph_vector(kernel=ker_m)
app_builder(x_in_1, x_in_2, x_out)
```

I am getting **[None, None]** as output i.e, the output from the kernel which should be [1] and [2]. Why I am getting None?
Also, when I checked the graph with **app_builder.display()** I can see data getting into the kernel but nothing came out of it.

**Reply by mariodruiz** [reply]
@soumitraghosh388,

You are running the behavior model, 

```python
    def behavioralfx(self):
        pass
```

which does not implement what you need. I would suggest you build the application and run it. Please, check notebooks 4_* to see how to do this.

**Reply by soumitraghosh388** [reply]
As I understand, **behavioralfx** is used for checks in the python layer which imitates the behaviour of main kernel in cpp. But the cpp file is the main execution that generates data. If I don't define **behavioralfx** it generates error : 
`AttributeError: type object 'Master' has no attribute 'behavioralfx'`

Is there is a way to omit  **behavioralfx**, what it must contain?

As you suggested I tried to build the application using `app_builder.build(x_in_1, x_in_2, x_out)` but got error : 
`ValueError: AIE.core(%tile02) either has no inputs or output objfifos`

**Reply by soumitraghosh388** [reply]
The issue got resolved after changing **behavioralfx** to :

```
  def behavioralfx(self):
      self.out0.array = self.in0.array
      self.out1.array = self.in1.array
```

**Reply by mariodruiz** [reply]
Hi @soumitraghosh388, were you able to build this for the NPU?
Currently, we have a limitation of two inputs and one output or one input and two outputs from the IT to the CT.

**Reply by soumitraghosh388** [reply]
While building I am getting error (I am using 9 kernels in total, using MT as well for broadcast) :
`ValueError: Unable to place kernel nw_stream_kernel_4 - no more free CT tiles`

I could not understand : "Currently, we have a limitation of two inputs and one output or one input and two outputs from the IT to the CT." Are these input and output ports of kernels?
Also, I could not find document explaining Rialto framework in details. Can you refer me to those (if any)?

**Reply by mariodruiz** [reply]
> While building I am getting error (I am using 9 kernels in total, using MT as well for broadcast) :
ValueError: Unable to place kernel nw_stream_kernel_4 - no more free CT tiles

We only support one kernel per compute tile at the moment. I should be writing a post to show how to do kernel fusion so you can map multiple kernels in one compute tile. In the meantime, you can check how this is done here: https://github.com/AMDResearch/Riallto/blob/main/tests/test_kernel_fusion.py

**Reply by mariodruiz** [reply]
> "Currently, we have a limitation of two inputs and one output or one input and two outputs from the IT to the CT." Are these input and output ports of kernels?

No, this is a limitation of input and output buffers from system memory via to the interface tile to the compute tile. A kernel itself could have two inputs and two output buffers.

**Reply by mariodruiz** [reply]
> Also, I could not find document explaining Rialto framework in details. Can you refer me to those (if any)?

Section 4 notebooks describes the framework in detail, starting from writing your own compute kernels to defining the dataflow graph. 
Notebook [4_2](https://riallto.ai/notebooks/4_2_write_your_kernel.html) and [4_5](https://riallto.ai/notebooks/4_5_describe_an_application.html) describe the two parts in detail. 

You can find information about the package here: https://riallto.ai/modules.html
